File: Code_Till_Phase3/python/benchmark.py
"""
================================================================================
  Python Imperative Comparison: Parallel Algorithms
================================================================================

PURPOSE:
  This file implements the same four computational problems in Python to serve
  as a second IMPERATIVE BASELINE for comparison with Haskell implementations.

  Python adds an interesting dimension to the comparison:
  - GIL (Global Interpreter Lock) prevents true multi-threading for CPU-bound work
  - Must use multiprocessing (separate processes) for CPU parallelism
  - This means inter-process communication overhead (pickling/unpickling)
  - Shows WHY language-level parallelism support matters

KEY OBSERVATIONS:
  1. Python's GIL makes threading useless for CPU-bound parallelism
  2. multiprocessing.Pool adds significant overhead (process creation, IPC)
  3. Python is 10-100x slower than Haskell/C++ for CPU-bound work
  4. BUT: Python code is very readable — good for showing what we're computing

RUN:
  python3 benchmark.py

================================================================================
"""

import time
import random
import math
import sys
from multiprocessing import Pool, cpu_count
from concurrent.futures import ProcessPoolExecutor, ThreadPoolExecutor


# ============================================================================
# Timing utility
# ============================================================================

def time_it(func, *args):
    """
    Measure wall-clock time of a function call.
    Returns (result, elapsed_ms).
    """
    start = time.perf_counter()
    result = func(*args)
    elapsed = (time.perf_counter() - start) * 1000  # ms
    return result, elapsed


def print_header(title):
    print()
    print("=" * 70)
    print(f"  {title}")
    print("=" * 70)
    print()


def print_result(label, time_ms):
    if time_ms < 1.0:
        print(f"  {label:<40}{time_ms*1000:.0f} us")
    elif time_ms < 1000:
        print(f"  {label:<40}{time_ms:.2f} ms")
    else:
        print(f"  {label:<40}{time_ms/1000:.3f} s")


# ============================================================================
# PROBLEM 1: Merge Sort
# ============================================================================

def merge(left, right):
    """
    Merge two sorted lists into one sorted list.
    
    COMPARISON WITH HASKELL:
      Haskell: merge [] ys = ys; merge xs [] = xs; merge (x:xs) (y:ys) = ...
      Python: Imperative with explicit indices and list building
      
    Both create a new list (neither is truly in-place).
    """
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result


def sequential_merge_sort(arr):
    """
    Sequential merge sort.
    
    COMPARISON WITH HASKELL:
      Very similar structure! Python's slice notation makes it almost as clean:
        Haskell: let (left, right) = splitAt mid xs
        Python:  left, right = arr[:mid], arr[mid:]
    """
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    left = sequential_merge_sort(arr[:mid])
    right = sequential_merge_sort(arr[mid:])
    return merge(left, right)


def _sort_chunk(arr):
    """Worker function for parallel sort (must be top-level for pickling)."""
    return sequential_merge_sort(arr)


def parallel_merge_sort(arr, num_workers=4):
    """
    Parallel merge sort using multiprocessing.
    
    CRITICAL DIFFERENCE FROM HASKELL:
      Python's GIL prevents true thread-level parallelism.
      We MUST use multiprocessing, which means:
      1. Each worker is a SEPARATE PROCESS (heavy to create)
      2. Data must be SERIALIZED (pickled) to send to workers
      3. Results must be DESERIALIZED when received back
      4. This IPC overhead can dominate for small inputs
      
      Haskell:
        - Sparks are nearly free (just a pointer on a work queue)
        - Data sharing via immutable memory (no copying needed)
        - GC handles cleanup automatically
    
    STRATEGY:
      1. Split array into num_workers chunks
      2. Sort each chunk in a separate process
      3. Merge sorted chunks pairwise
    """
    if len(arr) <= 1000:  # Don't parallelize small arrays
        return sequential_merge_sort(arr)
    
    chunk_size = max(1, len(arr) // num_workers)
    chunks = [arr[i:i+chunk_size] for i in range(0, len(arr), chunk_size)]
    
    with Pool(num_workers) as pool:
        sorted_chunks = pool.map(_sort_chunk, chunks)
    
    # Merge sorted chunks pairwise
    while len(sorted_chunks) > 1:
        merged = []
        for i in range(0, len(sorted_chunks), 2):
            if i + 1 < len(sorted_chunks):
                merged.append(merge(sorted_chunks[i], sorted_chunks[i+1]))
            else:
                merged.append(sorted_chunks[i])
        sorted_chunks = merged
    
    return sorted_chunks[0]


# ============================================================================
# PROBLEM 2: Matrix Multiplication
# ============================================================================

def sequential_mat_mul(a, b):
    """
    Sequential matrix multiplication using nested loops.
    
    COMPARISON WITH HASKELL:
      Haskell: [[ sum $ zipWith (*) row col | col <- transpose b ] | row <- a]
      Python:  Three nested for-loops (C-style iteration)
      
    The Haskell version is more declarative and closer to mathematical notation.
    """
    n = len(a)
    c = [[0.0] * n for _ in range(n)]
    for i in range(n):
        for k in range(n):
            for j in range(n):
                c[i][j] += a[i][k] * b[k][j]
    return c


def _compute_rows(args):
    """Worker: compute a chunk of result rows."""
    a_chunk, b, n = args
    result = []
    for row in a_chunk:
        new_row = [0.0] * n
        for k in range(n):
            for j in range(n):
                new_row[j] += row[k] * b[k][j]
        result.append(new_row)
    return result


def parallel_mat_mul(a, b, num_workers=4):
    """
    Parallel matrix multiplication using multiprocessing.
    
    Each worker computes a chunk of rows.
    
    COMPARISON WITH HASKELL:
      Haskell: parMap rdeepseq computeRow a  ← one line!
      Python:  Split data, pickle, send to processes, unpickle, collect, merge
    """
    n = len(a)
    chunk_size = max(1, n // num_workers)
    chunks = [(a[i:i+chunk_size], b, n) for i in range(0, n, chunk_size)]
    
    with Pool(num_workers) as pool:
        results = pool.map(_compute_rows, chunks)
    
    return [row for chunk in results for row in chunk]

# ============================================================================
# Data Generation
# ============================================================================

def generate_random_list(n, seed=42):
    rng = random.Random(seed)
    return [rng.randint(1, n * 10) for _ in range(n)]


def generate_random_matrix(n, seed=42):
    rng = random.Random(seed)
    return [[rng.uniform(0, 100) for _ in range(n)] for _ in range(n)]

# ============================================================================
# MAIN
# ============================================================================

def main():
    print()
    print("╔══════════════════════════════════════════════════════════════════╗")
    print("║   Python Imperative Comparison Benchmarks                      ║")
    print("║   NOTE: Uses multiprocessing (not threading) due to GIL        ║")
    print("╚══════════════════════════════════════════════════════════════════╝")
    
    # ===== PROBLEM 1: Merge Sort =====
    print_header("BENCHMARK 1: Merge Sort (multiprocessing.Pool)")
    
    for size in [10000, 50000, 100000]:
        print(f"  --- Input size: {size} elements ---")
        arr = generate_random_list(size)
        
        seq_result, seq_time = time_it(sequential_merge_sort, arr)
        print_result("Sequential Merge Sort", seq_time)
        
        for workers in [2, 4]:
            par_result, par_time = time_it(parallel_merge_sort, arr, workers)
            speedup = seq_time / par_time if par_time > 0 else 0
            print_result(f"Parallel ({workers} workers)", par_time)
            print(f"    Speedup: {speedup:.2f}x")
            correct = seq_result == par_result
            print(f"    Correctness: {'PASS' if correct else 'FAIL'}")
        print()
    
    # ===== PROBLEM 2: Matrix Multiplication =====
    print_header("BENCHMARK 2: Matrix Multiplication (multiprocessing.Pool)")
    
    for size in [64, 128]:
        print(f"  --- Matrix size: {size}x{size} ---")
        a = generate_random_matrix(size, 42)
        b = generate_random_matrix(size, 99)
        
        _, seq_time = time_it(sequential_mat_mul, a, b)
        print_result("Sequential", seq_time)
        
        for workers in [2, 4]:
            _, par_time = time_it(parallel_mat_mul, a, b, workers)
            speedup = seq_time / par_time if par_time > 0 else 0
            print_result(f"Parallel ({workers} workers)", par_time)
            print(f"    Speedup: {speedup:.2f}x")
        print()
    

    
    print("\nAll Python benchmarks complete!")


if __name__ == "__main__":
    main()

File: Code_Till_Phase3/cpp/benchmark.cpp
/*
================================================================================
  C++ Imperative Comparison: Parallel Algorithms
================================================================================

PURPOSE:
  This file implements the same four computational problems in C++ to serve
  as an IMPERATIVE BASELINE for comparison with our Haskell implementations.

  The key differences to observe are:
  1. SHARED MUTABLE STATE: C++ threads share memory, requiring explicit
     synchronization (mutexes, atomics) to prevent data races.
  2. MANUAL THREAD MANAGEMENT: We must create, join, and manage threads.
  3. IN-PLACE MUTATION: C++ sorts and modifies arrays in place.
  4. EXPLICIT SYNCHRONIZATION: Locks, condition variables, atomics.

  These are the "pain points" that FP's purity eliminates.

COMPILE:
  g++ -std=c++17 -O2 -pthread -o benchmark benchmark.cpp

RUN:
  ./benchmark
================================================================================
*/

#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <thread>
#include <future>
#include <random>
#include <mutex>
#include <map>
#include <string>
#include <sstream>
#include <numeric>
#include <functional>
#include <cmath>
#include <fstream>
#include <iomanip>
#include <atomic>
#include <cctype>

using namespace std;
using namespace chrono;

// ============================================================================
// Timing utility
// ============================================================================

template <typename Func>
double timeIt(Func f) {
    auto start = high_resolution_clock::now();
    f();
    auto end = high_resolution_clock::now();
    return duration_cast<microseconds>(end - start).count() / 1000.0; // ms
}

// ============================================================================
// PROBLEM 1: Merge Sort
// ============================================================================

/*
  IMPERATIVE MERGE SORT (Sequential)
  
  KEY DIFFERENCES FROM HASKELL:
  - IN-PLACE: We modify the array directly (no new allocations per level)
  - MUTABLE: The 'temp' buffer is reused across merge operations
  - INDEX-BASED: We use array indices, not recursive list splitting
  
  This is more memory-efficient than the Haskell version but:
  - Harder to reason about correctness (aliasing, buffer overflows)
  - Cannot safely share arrays between threads without synchronization
  - Must carefully manage the temp buffer's lifecycle
*/
void merge(vector<int>& arr, int left, int mid, int right, vector<int>& temp) {
    int i = left, j = mid + 1, k = left;
    while (i <= mid && j <= right) {
        if (arr[i] <= arr[j]) temp[k++] = arr[i++];
        else temp[k++] = arr[j++];
    }
    while (i <= mid) temp[k++] = arr[i++];
    while (j <= right) temp[k++] = arr[j++];
    for (int i = left; i <= right; i++) arr[i] = temp[i];
}

void sequentialMergeSort(vector<int>& arr, int left, int right, vector<int>& temp) {
    if (left >= right) return;
    int mid = left + (right - left) / 2;
    sequentialMergeSort(arr, left, mid, temp);
    sequentialMergeSort(arr, mid + 1, right, temp);
    merge(arr, left, mid, right, temp);
}

/*
  IMPERATIVE PARALLEL MERGE SORT
  
  KEY DIFFERENCES FROM HASKELL:
  - Uses std::async to spawn threads (similar to Haskell's 'async')
  - SHARED MUTABLE STATE: Both threads write to the SAME array
    (safe here because they write to different index ranges)
  - Must explicitly manage thread lifecycle (future.get() = join)
  - Depth threshold to avoid thread explosion (same concept as Haskell)
  
  DANGER ZONES:
  - If we mess up index ranges, we get data races → undefined behavior
  - No type-system protection against shared mutable access
  - Must remember to call .get() on futures (else threads are detached)
  
  CONTRAST WITH HASKELL:
  - Haskell's rpar creates lightweight sparks, not OS threads
  - Haskell's purity guarantees no data races AT COMPILE TIME
  - C++ requires the programmer to manually verify thread safety
*/
void parallelMergeSort(vector<int>& arr, int left, int right, vector<int>& temp, int depth) {
    if (left >= right) return;
    if (depth <= 0) {
        sequentialMergeSort(arr, left, right, temp);
        return;
    }
    int mid = left + (right - left) / 2;
    
    // Spawn left half on a new thread
    // NOTE: This shares 'arr' and 'temp' with the parent thread!
    // Safe ONLY because we write to non-overlapping index ranges.
    auto leftFuture = async(launch::async, [&]() {
        parallelMergeSort(arr, left, mid, temp, depth - 1);
    });
    
    // Right half on current thread
    parallelMergeSort(arr, mid + 1, right, temp, depth - 1);
    
    // Wait for left half to complete
    leftFuture.get();
    
    // Merge (sequential — both halves are now sorted)
    merge(arr, left, mid, right, temp);
}

// ============================================================================
// PROBLEM 2: Matrix Multiplication
// ============================================================================

/*
  IMPERATIVE MATRIX MULTIPLICATION (Sequential)
  
  Three nested for-loops — the classic O(n³) algorithm.
  
  KEY DIFFERENCES FROM HASKELL:
  - IN-PLACE: Result matrix is pre-allocated and filled
  - INDEX-BASED: No list comprehensions, just index arithmetic
  - CACHE-FRIENDLY: Row-major access pattern is efficient
  
  The C++ version is MUCH faster than Haskell's list-based version
  because arrays have better cache locality than linked lists.
*/
vector<vector<double>> sequentialMatMul(const vector<vector<double>>& a,
                                        const vector<vector<double>>& b) {
    int n = a.size();
    vector<vector<double>> c(n, vector<double>(n, 0.0));
    for (int i = 0; i < n; i++)
        for (int k = 0; k < n; k++)
            for (int j = 0; j < n; j++)
                c[i][j] += a[i][k] * b[k][j];
    return c;
}

/*
  IMPERATIVE PARALLEL MATRIX MULTIPLICATION
  
  Uses std::thread to parallelize the outer loop.
  Each thread computes a chunk of rows.
  
  KEY DIFFERENCES FROM HASKELL:
  - Haskell: parMap rdeepseq computeRow a (one line change!)
  - C++: Must manually partition rows, create threads, join them
  - C++: Threads share the output matrix (safe: non-overlapping writes)
  - C++: Must carefully compute chunk boundaries
*/
vector<vector<double>> parallelMatMul(const vector<vector<double>>& a,
                                       const vector<vector<double>>& b,
                                       int numThreads) {
    int n = a.size();
    vector<vector<double>> c(n, vector<double>(n, 0.0));
    
    vector<thread> threads;
    int chunkSize = (n + numThreads - 1) / numThreads;
    
    for (int t = 0; t < numThreads; t++) {
        int startRow = t * chunkSize;
        int endRow = min(startRow + chunkSize, n);
        
        threads.emplace_back([&, startRow, endRow]() {
            for (int i = startRow; i < endRow; i++)
                for (int k = 0; k < n; k++)
                    for (int j = 0; j < n; j++)
                        c[i][j] += a[i][k] * b[k][j];
        });
    }
    
    for (auto& t : threads) t.join();
    return c;
}

// ============================================================================
// Random Data Generation
// ============================================================================

vector<int> generateRandomList(int n, int seed) {
    mt19937 gen(seed);
    uniform_int_distribution<int> dist(1, n * 10);
    vector<int> v(n);
    for (int i = 0; i < n; i++) v[i] = dist(gen);
    return v;
}

vector<vector<double>> generateRandomMatrix(int n, int seed) {
    mt19937 gen(seed);
    uniform_real_distribution<double> dist(0.0, 100.0);
    vector<vector<double>> m(n, vector<double>(n));
    for (int i = 0; i < n; i++)
        for (int j = 0; j < n; j++)
            m[i][j] = dist(gen);
    return m;
}


// ============================================================================
// Helper: Print formatted result
// ============================================================================

void printResult(const string& label, double timeMs) {
    cout << "  " << left << setw(40) << label;
    if (timeMs < 1.0) cout << fixed << setprecision(0) << timeMs * 1000 << "us" << endl;
    else if (timeMs < 1000.0) cout << fixed << setprecision(2) << timeMs << " ms" << endl;
    else cout << fixed << setprecision(3) << timeMs / 1000.0 << " s" << endl;
}

// ============================================================================
// MAIN: Run all benchmarks
// ============================================================================

int main() {
    cout << endl;
    cout << "╔══════════════════════════════════════════════════════════════════╗" << endl;
    cout << "║   C++ Imperative Comparison Benchmarks                         ║" << endl;
    cout << "╚══════════════════════════════════════════════════════════════════╝" << endl;
    
    // ===== PROBLEM 1: Merge Sort =====
    cout << "\n======================================================================" << endl;
    cout << "  BENCHMARK 1: Merge Sort (std::async / std::thread)" << endl;
    cout << "======================================================================\n" << endl;
    
    for (int size : {10000, 50000, 100000, 500000}) {
        cout << "  --- Input size: " << size << " elements ---" << endl;
        auto original = generateRandomList(size, 42);
        
        { // Sequential
            auto arr = original;
            vector<int> temp(size);
            double t = timeIt([&]() {
                sequentialMergeSort(arr, 0, size - 1, temp);
            });
            printResult("Sequential Merge Sort", t);
            
            // Parallel
            for (int depth : {2, 3, 4}) {
                auto arr2 = original;
                vector<int> temp2(size);
                double pt = timeIt([&]() {
                    parallelMergeSort(arr2, 0, size - 1, temp2, depth);
                });
                string label = "Parallel (depth=" + to_string(depth) + ")";
                printResult(label, pt);
                cout << "    Speedup: " << fixed << setprecision(2) << t / pt << "x" << endl;
                
                // Verify correctness
                bool correct = (arr == arr2);
                cout << "    Correctness: " << (correct ? "PASS" : "FAIL") << endl;
            }
        }
        cout << endl;
    }
    
    // ===== PROBLEM 2: Matrix Multiplication =====
    cout << "======================================================================" << endl;
    cout << "  BENCHMARK 2: Matrix Multiplication (std::thread)" << endl;
    cout << "======================================================================\n" << endl;
    
    for (int size : {64, 128, 256}) {
        cout << "  --- Matrix size: " << size << "x" << size << " ---" << endl;
        auto a = generateRandomMatrix(size, 42);
        auto b = generateRandomMatrix(size, 99);
        
        double seqTime;
        {
            seqTime = timeIt([&]() { sequentialMatMul(a, b); });
            printResult("Sequential", seqTime);
        }
        
        for (int threads : {2, 4, 8}) {
            double pt = timeIt([&]() { parallelMatMul(a, b, threads); });
            string label = "Parallel (" + to_string(threads) + " threads)";
            printResult(label, pt);
            cout << "    Speedup: " << fixed << setprecision(2) << seqTime / pt << "x" << endl;
        }
        cout << endl;
    }

    
    cout << "\nAll C++ benchmarks complete!" << endl;
    return 0;
}
